### ISSUE TITLE
Simplify extensions that provide commands and how those commands are integrating into formatting and saving files

### ISSUE THREAD

**Issue by julie777** [microsoft/vscode#145515]
There are so many related issues, and I have created some of them, that I would like to propose a simplification that will allow extensions and users more control and easier use of tools that manipulate files. (linters, formatters, etc.)

# Goal
The goal is to provide:
- easy discovery of the commands that can be used to format, lint, etc. for a given file type
- user control over which commands run automatically, when and in what order they run
- a way for a user to debug/review the actual commands that ran and whether they succeeded or failed.

Note: One question I have is about "language." vs "editor.". Is "editor." equivalent to "any language not specifying this setting"?
I am going to assume that for my suggestion.

## Reference
Using this extract of a settings file to work from,  and I tried to limit it to the kinds of things that can be related to formatting and saving files:
```  
  "files.eol": "\n",
  "python.formatting.provider": "autopep8",
  "python.linting.enabled": true,
  "python.linting.flake8Enabled": true,
  "python.linting.pylintEnabled": true,
  "[python]": {
   "editor.formatOnSave": true
  },
  "editor.defaultFormatter": "ms-python.python",
  "editor.codeActionsOnSave": {
     "source.organizeImports": true,
  },
  "editor.formatOnSave": true,
  "files.trimFinalNewlines": true,
```
## Change the way settings work for formatting and run on save

All of the settings could be applied to all files, to a specific type of file and to a language setting for files. (This in another issue ??)
This makes it very hard to determine what final value for a setting is being used.

Change the settings to:
```
// this allows it to be type of file specific if desired. also , I don't understand why file and editor are separated
"editor.eol": "\n"         
// this was trimFinalNewlines but it really should get rid of all tailing white space and leave a single newline folowing the last line of text
"editor.trimEndingWhiteSpace   
// I'm not clear about using editor for actions as they will vary per file type so this only controls file types with a formatter defined.
"editor.formatOnSave": True         
"python.formatters": ['autopep8", "isort"]
"python.linters": ["flake8", "pylint"]
// can be set generally, using the general types that will in turn use specific commands for the file type, or it can be overridden by file type
"editor.runOnSave": [
        // forces all formatters to run based on file type.
       "format",  
       // forces all defined linters for a file type to run when saving the file, can be overridden by file type
       "lint",          
       "mycommand",
"python.runOnSave": ["black", "pylint", "mccabe"]
]
```
I got ride of editor.default formatter, as there is no universal formatter for all file types.
Get rid of "source.organizeImports" because it requires specificity that my solutions don't need.

# Description
The editor settings are universal unless overridden, but contingent on available providers.
## Formatting
1. settings has a list of available formatters for each file type, each formatter has a format command in the commands menu
      - individual commands can be run from the format menu
      - the general `format` command is supplied by vscode runs all selected formatters for the file type
2. the user can choose one or more of the available formatters and add them to the formatters list for a given file type
    - the user sets the order if selecting multiple formatters
    - the general `format` command will run them in order
    - the general `format` command can be invoked by a shortcut, from the command menu, or from any context menu where it appears
    - individual format commands or the general `format` command are available to put in the list of actions on save
## Run on Save
1. all commands are available to be put in the runOnSave list for each file type
2. the user can put any (reasonable) command on the list in the order they choose and the commands will run in that order

## Extensions
I haven't developed extensions and I am not that knowlegeable on the topic, so my thoughts are based on a vscode user's perspective.
One thing extensions do is provide commands and that is all I am covering here.
1. extensions provide commands and for each command one or more usage types (formatter, onSave, linter, etc.) and file types that the command works on (.js, .ts, etc.).
2. settings can use usage types and file types to provider the user an appropriate list of possibility for formatters for a given file type or commands that can be used for run on save for the file type.
3. The user is in control of which ones get used and in what way.

## Logging
In order for users to understand what is really happening, and to allow them to see if the wrong commands are running, if there are errors, etc. all commands must put an entry in the extension host log when run.

When running commands
- they should provide an entry in the extension host log to show that they have started running. 
- log any stderr output to the extension host log
- record the result code from execution in the extension host log.

# Summary
The above solution generalizes formatting and run on save, and puts the vscode user in control. It makes it easy to discover what runs and when. It provides a simple and explainable way for the user to take control.

**Comment by julie777** [comment]
Today I finally found the error output from trying to run autopep8 in the output pane after selecting Python in the dropdown. I'm not sure why I didn't see that before. Maybe I was too focused looking at the log choices.

I will give you an example of the kind of problems I have been having.
I finally got both formatting and sorting imports (formatting) working. It was easier once I found the output from running actions on save.
```
    "[python]": {
    "formatOnSave": true,
    
    "defaultFormatter": "ms-python.python",
        "formatting.provider": "autopep8",
    }
   "editor.codeActionsOnSave": {
         "source.organizeImports": true,
   },
```
Without the default formatter set to ms-python.python (an extension) it won't work. I finally found this sentence in defaultsettings
         "Defines a default formatter which takes precedence over all other formatter settings."

It had previously been set on editor, but that won't work if I have non python files that need a different default formatter.
I moved it into a Python language block and that worked.

However, when I moved codeActionsOnSave to the Python block it didn't work.

I haven't found any place in the documentation that tells me which settings are language settings and which aren't.
I already figured out that "python.editor.tabSize" doesn't work because tabSize is a language setting and python.x.y is not a language setting even though it starts with a language.

There are several problems that interact.
- you can't just look at settings and see what value is going to be used
- there is no place that lists all valid [language] settings
- the complicated interaction of codeActionsOnSave and format on save and lint on save get very messy.

It would be so much easier if I could just put
```
   "[python]": {
        "codeActionsOnSave": [
            "isort",
            "autopep8",
           "flake8",
    }
```
And know that is exactly what will happen when I save a python file.

**Comment by julie777** [comment]
And yet another piece of the puzzle. Previously when working on a Quasar/Vue/Javascript project this ended up in my user settings.
```
  
  "editor.codeActionsOnSave": {
    "source.organizeImports": true,
    "attr-sorter": true,
    "source.fixAll.eslint": true
  }, 
```

So, codeActionsOnSave can't be specified as a language setting, so some extensions provide actions in editorActionsOnSave which are certainly file type/language specific.  attr-sorter is for html and source.fixAll.eslint is for javascript.

I really hope that I am demonstrating how hard it is for a vscode user who works on multiple projects in multiple languages to get everything to work "just right".

**Comment by jrieken** [comment]
> Note: One question I have is about "language." vs "editor.". Is "editor." equivalent to "any language not specifying this setting"?

I can answer that question. The rest of this issue seems like it mixing many things together and don't really know what to do with that.

Generally there are editor features (or feature skeletons) and language extensions bring these features to live. E.g there `editor.formatOnSave` but the editor doesn't actually know how to format anything. For that it needs extensions, like an extension for TypeScript or Python. These extension register as "feature provider" and that usually comes with an extra set of settings, like what formatting flavour to use or how to do what. 

The same is true for code actions. They are similar to formatting but solve a different problem, like "help me refactor this code" or "help me fix this error". Again, VS Code is just the shell and extensions provide the actual implementations. 

Some of these features, like formatting and code actions, can run in various ways. Like format a document manually via "F1 > Format Document" or format automatically when saving a file. That's why `formatOnSave` and `codeActionsOnSave` exists, both can be configure per language, and the detail differ a little. There can only be on formatter (hence the default formatter setting) but there can be multiple code actions that run in sequence (there is also a related issue to allow multiple sequential formatters to run)

**Comment by jrieken** [comment]
I hope that I could clarify some confusion. I see how things are very intuitive but there also things that aren't under our control, like what settings a contributed formatter or code action provider accepts
